src/py/flwr/common/auth_plugin/keycloak_plugin.py

```diff
@@ -0,0 +1,271 @@
+
+        payload = {
+            "client_id": self.keycloak_client_id,
+            "client_secret": self.keycloak_client_secret,
+            "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
+            "device_code": device_code,
+        }
+
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
+
+        response = post(self.token_url, data=payload, headers=headers, timeout=10)
+        if response.status_code == 200:
+            data = response.json()
+            access_token = data.get("access_token")
+            refresh_token = data.get("refresh_token")
+            auth_tokens = {
+                "access_token": access_token,
+                "refresh_token": refresh_token,
+            }
+            return GetAuthTokenResponse(auth_tokens=auth_tokens)
+
+        return GetAuthTokenResponse(auth_tokens={})
+
+    def refresh_token(self, context: grpc.ServicerContext) -> bool:
+        """Refresh auth tokens in the provided metadata."""
+        metadata = context.invocation_metadata()
+        refresh_token = _get_value_from_tuples(_REFRESH_TOKEN_HEADER, metadata)
+        if not refresh_token:
+            return False
+
+        payload = {
+            "client_id": self.keycloak_client_id,
+            "client_secret": self.keycloak_client_secret,
+            "grant_type": "refresh_token",
+            "refresh_token": refresh_token.decode("utf-8"),
+        }
+
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
+
+        response = post(self.token_url, data=payload, headers=headers, timeout=10)
+        if response.status_code == 200:
+
+            data = response.json()
+            access_token = data.get("access_token")
+            refresh_token = data.get("refresh_token")
+
+            context.send_initial_metadata(
+                (
+                    (
+                        _AUTH_TOKEN_HEADER,
+                        access_token,
+                    ),
+                    (
+                        _REFRESH_TOKEN_HEADER,
+                        refresh_token,
+                    ),
+                )
+            )
+            return True
+
+        return False
+
+
+class KeycloakUserPlugin(UserAuthPlugin):
+    """Abstract Flower User Auth Plugin class."""
+
+    def __init__(self, config: dict[str, Any], config_path: Path):
+        self.access_token = config["access-token"]
+        self.refresh_token = config["refresh-token"]
+        self.config = config
+        self.config_path = config_path
+
+    @staticmethod
+    def login(
+        login_details: dict[str, str],
+        config: dict[str, Any],
+        federation: str,
+        exec_stub: ExecStub,
+    ) -> dict[str, Any]:
+        """Read relevant auth details from federation config."""
+        timeout = int(login_details.get("expires_in", "600"))
+        interval = int(login_details.get("interval", "5"))
+        device_code = login_details.get("device_code")
```

Comment:
Could we create a new python file, say `keycloak_constant.py`, and define these keys as constants, like [here](https://github.com/adap/flower/blob/main/src/py/flwr/common/constant.py#L114)?